File: python/src/dynamic_range_minimum_queries.py
import sys
import math

class SegmentTree:

    @staticmethod
    def parent(i):
        return (i - 1) // 2

    @staticmethod
    def left(i):
        return i * 2 + 1

    @staticmethod
    def right(i):
        return i * 2 + 2

    @staticmethod
    def sibling(i):
        if i % 2 == 0:
            return i - 1
        else:
            return i + 1

    # should only be called using an index from the last level
    def original_index(self, tree_index):
        return tree_index - (2**(self.n - 1) - 1)

    # should only be called using an index within bounds of the original array
    def tree_index(self, original_index):
        return original_index + (2**(self.n - 1) - 1)

    def comparison(self, i, j):
        return self.tree[i] < self.tree[j]

    def __init__(self, values) -> None:
        self.n = math.ceil(math.log2(len(values))) + 1
        self.tree = [0 for i in range(2**self.n - 1)]

        # fill initial values
        for (i, value) in zip(range(2**(self.n - 1) - 1, 2**self.n - 1), values):
            self.tree[i] = value

        for i in range(2**self.n - 1 - 1, 0, -2):
            if self.comparison(i, i - 1):
                self.tree[self.parent(i)] = self.tree[i]
            else:
                self.tree[self.parent(i)] = self.tree[i - 1]

    def search(self, value):
        index = 0

        if value > self.tree[index]:
            return None

        while index < (2**(self.n - 1) - 1):

            if value <= self.tree[self.left(index)]:
                index = self.left(index)
                continue

            if value <= self.tree[self.right(index)]:
                index = self.right(index)
                continue

        return (index, self.tree[index])

    def update_value(self, tree_index, value):
        self.tree[tree_index] = value

        while tree_index != 0:
            sibling_index = self.sibling(tree_index)

            if self.comparison(tree_index, sibling_index):
                self.tree[self.parent(tree_index)] = self.tree[tree_index]
            else:
                self.tree[self.parent(tree_index)] = self.tree[sibling_index]

            tree_index = self.parent(tree_index)

    def return_original_values(self):
        return self.tree[2**(self.n - 1) - 1, 2**self.n - 1]

    def min_range_query(self, a, b, left_boundary, right_boundary, index):
        mid_way = left_boundary + (right_boundary - left_boundary) // 2

        if a <= left_boundary and right_boundary <= b:
            return self.tree[index]

        elif right_boundary < a or b < left_boundary:
            return sys.maxsize

        else:
            return min(
                self.min_range_query(a, b, left_boundary, mid_way, self.left(index)),
                self.min_range_query(a, b, mid_way + 1, right_boundary, self.right(index))
            )




if __name__ == "__main__":

    first_line = sys.stdin.readline().split(" ")
    n_numbers = int(first_line[0])
    q_queries = int(first_line[1])

    values = list(map(int, sys.stdin.readline().split(" ")))
    tree = SegmentTree(values)

    for q in range(q_queries):

        query = list(map(int, sys.stdin.readline().split(" ")))

        # update query
        if query[0] == 1:
            index = query[1] - 1
            value = query[2]
            tree.update_value(tree.tree_index(index), value)

        # look up query
        else:
            a = query[1] - 1
            b = query[2] - 1
            print(f"{tree.min_range_query(a, b, 0, n_numbers - 1, 0)}")

File: rust/src/bin/dynamic_range_minimum_queries.rs
use std::{
    cmp::min,
    io::{stdin, Read},
};

struct SegmentTree {
    array_len: usize,
    tree: Vec<usize>,
}

impl SegmentTree {
    #[inline]
    // 1 indexed segment tree
    fn left(i: usize) -> usize {
        i * 2
    }

    #[inline]
    // 1 indexed segment tree
    fn right(i: usize) -> usize {
        i * 2 + 1
    }

    #[inline]
    fn parent(i: usize) -> usize {
        i / 2
    }

    fn new(values: Vec<usize>) -> Self {
        let array_len = values.len();
        let tree_len = array_len * 2;
        let mut tree = vec![usize::MAX; tree_len];

        tree[array_len..].copy_from_slice(&values);

        let mut segment_tree = SegmentTree { array_len, tree };

        segment_tree.treeify();
        segment_tree
    }

    fn treeify(&mut self) {
        for i in (1..self.array_len).into_iter().rev() {
            self.tree[i] = min(
                self.tree[SegmentTree::left(i)],
                self.tree[SegmentTree::right(i)],
            );
        }
    }

    // 0 indexed update on original array element
    fn update_value(&mut self, index: usize, value: usize) {
        let mut tree_index = self.array_len + index;
        self.tree[tree_index] = value;

        tree_index = SegmentTree::parent(tree_index);

        while tree_index != 0 {
            self.tree[tree_index] = min(
                self.tree[SegmentTree::left(tree_index)],
                self.tree[SegmentTree::right(tree_index)],
            );
            tree_index = tree_index / 2;
        }
    }

    // query for range [l,r) i.e. left inclusive, right exclusive
    // 1 indexed query
    fn find_for_range(&self, query_left: usize, query_right: usize) -> usize {
        let mut value = usize::MAX;
        let mut l = query_left + self.array_len;
        let mut r = query_right + self.array_len;

        // the algorithm works by only adding values
        // that are add odd indices of the 1 indexed
        // segment tree array
        while l < r {
            // if l is odd then it is the right child
            // of it's parent so it can be added as is
            // incrementing l brings it to the next pair
            // dividing by 2 makes it the parent of the
            // next pair
            // if l is even then it is the left child
            // of the pair and the sum of the pair can
            // be found at the parent unless the interval
            // is closed by the right border
            if l % 2 != 0 {
                value = min(self.tree[l], value);
            }

            if r % 2 != 0 {
                r -= 1;
                value = min(self.tree[r], value);
            }

            l += 1;
            l /= 2;
            r /= 2;
        }

        value
    }
}

fn main() {
    let mut input = String::new();
    stdin().read_to_string(&mut input).unwrap();
    let mut input = input.lines();

    let first_line: Vec<usize> = input
        .next()
        .unwrap()
        .split_whitespace()
        .map(|val| val.parse().unwrap())
        .collect();
    let _n_numbers: usize = first_line[0];
    let q_queries: usize = first_line[1];

    let numbers: Vec<usize> = input
        .next()
        .unwrap()
        .split_whitespace()
        .map(|value| value.parse().unwrap())
        .collect();

    let mut tree = SegmentTree::new(numbers);

    for _ in 0..q_queries {
        let query: Vec<usize> = input
            .next()
            .unwrap()
            .split_whitespace()
            .map(|val| val.parse().unwrap())
            .collect();

        match query[0] {
            // update query
            1 => {
                // 0 indexed
                tree.update_value(query[1] - 1, query[2]);
            }
            // only other query is range query
            _ => {
                // 0 indexed
                println!("{}", tree.find_for_range(query[1] - 1, query[2]));
            }
        }
    }
}
